### PR TITLE
The adapter should support SampledFunctionParameterType to arrays mapping/transfer #62

### LINKED ISSUE #62: The adapter should support SampledFunctionParameterType to arrays mapping/transfer
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the DEHP-EcosimPro
- [x] I have searched [open](https://github.com/RHEAGROUP/DEHP-ECOSIMPRO/issues) and [closed](https://github.com/RHEAGROUP/DEHP-ECOSIMPRO/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

The  adapter currently doesn't let the user select the whole valueset set or multiple values from a SampledFunctionParameterType. In fact only one to one mapping is doable from Hub to Dst.

To overcome this, the adapter should let the user select a source the whole valueset from a SFPT or a part of it to map to a opc array variable type.

The way arrays are handled in a opc server is that for each array index one variable is there. 
Example: arrayvar[0,0],  arrayvar[0,1],  arrayvar[1,0],  arrayvar[1,1] ...

So by selecting any one of one opc array variable, the adapter should then automatically determine:

- Which variables are concerned (belongs to the same array)
- The dimension of the array
- Is the valueset or the selection fits

### Tasks
- [ ] Create a VariableRowVariable that represents a collection of variable. 
- [ ] Update the MappedElementDefinitionRowViewModel so it can store mapping of arrays ?
- [ ] Updates the Mapping configuration dialog
   - [ ] Allow mapping only dimensional compatible parameter
- [ ] Saving the mapping
- [ ] Ensuring that the loading of saved mapping works with mapping that maps SFPT to OPC Arrays 
- [ ] Tests

### System Configuration

- DEHPCommon version:
- Environment (Operating system, version and so on):
- .NET Framework version:
- Additional information:

### PR DESCRIPTION
**Pull request by arielleatrhea**
Created a VariableRowVariable that represents a collection of variable ;  Update the MappedElementDefinitionRowViewModel so it can store mapping of arrays ;  Updates the Mapping configuration dialog ; Allow mapping only dimensional compatible parameter ;  Saving the mapping ;  Ensuring that the loading of saved mapping works with mapping that maps SFPT to OPC Arrays ; Tests

fixes #62 

### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/DEHP-ECOSIMPRO/pulls) open
- [x] I have verified that I am following the DEHP-EcosimPro [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/DEHP-ECOSIMPRO/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
- [x]  Create a VariableRowVariable that represents a collection of variable.
- [x] Update the MappedElementDefinitionRowViewModel so it can store mapping of arrays ?
- [x] Updates the Mapping configuration dialog
- [x] Allow mapping only dimensional compatible parameter
- [x]  Saving the mapping
- [x]  Ensuring that the loading of saved mapping works with mapping that maps SFPT to OPC Arrays
- [x]  Tests

### CODE REVIEW

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 1]
```diff
@@ -53,13 +59,45 @@ namespace DEHPEcosimPro.ViewModel.Dialogs
     /// </summary>
     public class HubMappingConfigurationDialogViewModel : MappingConfigurationDialogViewModel, IHubMappingConfigurationDialogViewModel
     {
+        /// <summary>
+        /// The <see cref="INavigationService" />
+        /// </summary>
+        private readonly INavigationService navigation;
+        
+        /// <summary>
+        /// </summary>
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:68]: No docs for all the fields below

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 2]
```diff
@@ -53,13 +59,45 @@ namespace DEHPEcosimPro.ViewModel.Dialogs
...
+        public ChooseMappingColumnsViewModel ListOfRowsToMap { get; set; } = new ChooseMappingColumnsViewModel();
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:69]: Variable name says List but this is a viewmodel.

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 3]
```diff
@@ -164,6 +202,8 @@ public ElementDefinition SelectedElementDefinition
         /// </summary>
         private bool canContinue;
 
+        private IStatusBarControlViewModel statusBar;
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:205]: No docs

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 4]
```diff
@@ -180,9 +220,11 @@ public bool CanContinue
         /// <param name="dstController">The <see cref="IDstController"/></param>
         /// <param name="statusBar">The <see cref="IStatusBarControlViewModel"/></param>
         public HubMappingConfigurationDialogViewModel(IHubController hubController,
-            IDstController dstController, IStatusBarControlViewModel statusBar, IMappingConfigurationService mappingService) :
+            IDstController dstController, IStatusBarControlViewModel statusBar, IMappingConfigurationService mappingService, INavigationService navigation) :
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:223]: `mappingService` and `navigation` missing docs. Also `navigation` should be `navigationService`

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 5]
```diff
@@ -286,22 +326,56 @@ private void SelectedElementDefinitionChanged()
...
+                    userSelectecValue = this.navigation.ShowDxDialog<ChooseMappingColumns, ChooseMappingColumnsViewModel>(this.ListOfRowsToMap);
+
+                    if (userSelectecValue == true)
+                    {
+                        this.MapParameterToVariable(table.ListOfTuple, parameter);
+                    }
+                }
+                else
+                {
+                    //not compatible, show error to user
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:355]: ???

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 6]
```diff
@@ -390,19 +519,28 @@ private void SetSelectedMappedElement(Thing parameter)
...
+                foreach (var element in this.DstController.HubMapResult)
+                {
+                    element.VerifyValidity();
+                    this.MappedElements.Add(element);
+                }
+                this.CheckCanExecute();
             }
-
-            this.CheckCanExecute();
+            //foreach (var thing in this.ElementDefinitions.SelectMany(x => x.Parameter))
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:531]: Delete commented code.

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 7]
```diff
@@ -313,21 +387,76 @@ private void SelectedThingChanged()
...
+                        {
+                            this.MappedElements.Add(this.CreateMappedElement((ParameterOrOverrideBase) parameterOption.Thing));
+
+                            this.SelectedMappedElement = this.MappedElements
+                                .FirstOrDefault(x => x.SelectedParameter?.Iid == parameterOption.Thing.Iid);
+                        }
+
+                        if (this.SelectedAvailableVariables != null)
+                        {
+                            this.AreVariableTypesAreCompatible();
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:436]: Double `Are` in the name.

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 8]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
                     .Where(x => this.DstController.IsVariableWritable(x.Reference))
-                    .Select(r =>
+                    .ToList();
+
+            if (availableVariables.Count == 0)
+            {
+                return;
+            }
+
+            var arrayVariables = availableVariables.Where(x => x.Reference.DisplayName.Text.Contains("[")).ToList();
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:589]: Is there no better way to detect arrays?

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 9]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+
+            this.AvailableVariables.AddRange(availableVariables.Where(x => arrayVariables
+                    .All(y => y.Reference.NodeId.Identifier != x.Reference.NodeId.Identifier))
+                .Select(x =>
+                {
+                    var (reference, _) = x;
+                    return new VariableRowViewModel((reference, this.DstController.ReadNode(x.Reference)));
+                }));
+
+            var grouby = arrayVariables.GroupBy(x => this.GetNameFromArrayReference(x.Reference));
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:599]: Bad variable name.

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 10]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+                }
+            }
+
+            this.MappedElements.ChangeTrackingEnabled = true;
+
+            this.IsBusy = false;
+        }
+
+        /// <summary>
+        /// </summary>
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:631]: No docs

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 11]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+        public void SelectedAvailableVariablesChanged()
+        {
+            this.AreVariableTypesAreCompatible();
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="reference"></param>
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:639]: No docs

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 12]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+            if (splitedName.Length > 2)
+            {
+                return splitedName[0] + splitedName[2];
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        ///
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:659]: no docs

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 13]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+        /// </summary>
+        /// <param name="variable"></param>
+        /// <param name="parameter"></param>
+        /// <param name="compatibilityOfColumns"></param>
+        /// <returns></returns>
+        private bool AreArraysCompatible(VariableBaseRowViewModel variable, ParameterOrOverrideBase parameter)
+        {
+            var rowsAndColumnsOfVariable = variable.ActualValue.ToString().Split('[', 'x', ']');
+
+            var columnsOfParameter = parameter.ParameterType.NumberOfValues;
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:669]: I dont think this is correct. `NumberOfValues` will not give you number of columns of a SampledFunction. Please unit test this method thoroughly with a large number of test cases for different sizes (i.e. 1x1 , 1 x m, n x 1, a few m x n, n x n )

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 14]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="variable"></param>
+        /// <param name="parameter"></param>
+        /// <param name="compatibilityOfColumns"></param>
+        /// <returns></returns>
+        private bool AreArraysCompatible(VariableBaseRowViewModel variable, ParameterOrOverrideBase parameter)
+        {
+            var rowsAndColumnsOfVariable = variable.ActualValue.ToString().Split('[', 'x', ']');
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:667]: This is not the right way to determine sizes. You have to look at the actual underlying model elements.

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 15]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+
+            var columnsOfParameter = parameter.ParameterType.NumberOfValues;
+
+            var columnByRowsOfParameter = parameter.QueryParameterBaseValueSet
+                (this.SelectedOption, this.SelectedState).ActualValue.Count ;
+
+            string rowsOfVariable = "", columnsOfVariable = "";
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:674]: we always declare `var`.

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 16]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+            else
+            {
+                return false;
+            }
+
+            var isNumberOfRowsOfVariable = int.TryParse(rowsOfVariable, out var numberOfRowsOfVariable);
+
+            var numberOfRowsOfParameter = columnByRowsOfParameter / columnsOfParameter;
+
+            bool areRowsCompatibles = false ;
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:690]: same here. use `var`

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 17]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+            if (isNumberOfRowsOfVariable && numberOfRowsOfParameter != 0 && numberOfRowsOfVariable != 0)
+            {
+                areRowsCompatibles = numberOfRowsOfParameter == numberOfRowsOfVariable;
+            }
+
+            return areRowsCompatibles;
+        }
+
+        /// <summary>
+        ///
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:701]: no docs.

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 18]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+            var listOfValueByRowsAndColumns = this.DstController.Variables
+                .Where(x => this.DstController.IsVariableWritable(x.Reference)
+                            && this.GetNameFromArrayReference(x.Reference) == variableName).ToList();
+
+            var listOfVariableRowViewModel = listOfValueByRowsAndColumns.Select(x =>
+            {
+                var (reference, _) = x;
+                return new VariableRowViewModel((reference, this.DstController.ReadNode(x.Reference)));
+            }).ToList();
+            for (int i = 0; i < listOfVariableRowViewModel.Count(); i++)
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:719]: Add new line before the `for`.

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 19]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
 
-            this.MappedElements.ChangeTrackingEnabled = true;
+                        element.VerifyValidity();
+
+                        if (element.IsValid)
+                        {
+                            lisOfMappedElement.Add(element);
+                            this.MappedElements.Add(element);
+                        }
+
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:746]: unnecessary new line

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 20]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+                    }
+                    var result = lisOfMappedElement.Select(x => x.SelectedVariable.Name).ToList();
+                }
+
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:750]: unnecessary new line

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 21]
```diff
@@ -438,21 +576,213 @@ private void UpdateProperties()
...
+            }
+        }
+
+        /// <summary>
+        /// Get as a number the row and the colomn out of the name
+        /// </summary>
+        /// <param name="name">name of a value, should include [x] or [x,y]</param>
+        /// <param name="rowNumber">the row of the value</param>
+        /// <param name="colNumber">the column of the value</param>
+        private void GetRowAndColumnNumberFromName(string name, out int rowNumber, out int colNumber)
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:760]: It looks like there was another method above that was doing the same thing?

Either way, same question as before, is there no better way to find dimensions than from the string?

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+            this.ActualValue = this.GetDimension();
+            this.SetVariableRowName();
+        }
+
+        /// <summary>
+        /// list of <see cref="VariableRowViewModel" />
+        /// </summary>
+        public ReactiveList<VariableRowViewModel> Variables { get; set; } = new ReactiveList<VariableRowViewModel>();
+
+        public bool IsList { get; private set; }
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:72]: no docs

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+
+            return "/";
+        }
+
+        /// <summary>
+        /// If the Variable row is an array, return the dimension of the array
+        /// </summary>
+        /// <param name="arrayName">list of name of the data, should have [x] or [x,y]</param>
+        /// <param name="result">a string of the dimension, like [AxB]</param>
+        /// <returns></returns>
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:109]: no docs

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+        public ICloseWindowBehavior CloseWindowBehavior { get; set; }
+
+        /// <summary>
+        /// Get the column number of the data from it's name
+        /// </summary>
+        /// <param name="argName">name of the sata, should contain [x] of [x,y]</param>
+        /// <returns>as a string, a number or a comma then a number</returns>
+        private string GetColumnNumberFromValueName(string argName)
+        {
+            var probableindex = argName.Split('[');
```
Comment (alexatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs:125]: Again you are splitting strings, seems like you have done it in several places.

File: DEHPEcosimPro/Views/Dialogs/ChooseMappingColumns.xaml.cs [hunk 25]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
```
Comment (alexatrhea) [inline DEHPEcosimPro/Views/Dialogs/ChooseMappingColumns.xaml.cs:1]: Missing header

**Review by alexatrhea — CHANGES_REQUESTED** [review]

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+    {
+        /// <summary>
+        /// The represented <see cref="ReferenceDescription" /> corresponding <see cref="DataValue" />
+        /// </summary>
+        private readonly DataValue data;
+
+        /// <summary>
+        /// The represented <see cref="ReferenceDescription" />
+        /// </summary>
+        public readonly ReferenceDescription Reference;
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:47]: Hides inherited Reference, also why a public field?

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+
+        /// <summary>
+        /// Backing field for <see cref="Name" />
+        /// </summary>
+        private string name;
+
+        /// <summary>
+        /// Constructor of <see cref="ArrayVariableRowViewModel" />
+        /// </summary>
+        /// <param name="key">indice of the array</param>
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:57]: Is it really the indices of the array or the dimensions or something else?

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs (hunk 22 above)
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:72]: missing doc

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs (hunk 23 above)
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:109]: missing doc

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+
+        /// <summary>
+        /// Get the dimension of the array
+        /// </summary>
+        /// <returns>string presented like [AxB]</returns>
+        private string GetDimension()
+        {
+            var arrayName = this.Variables.Select(x => x.Name);
+
+            if (arrayName != null)
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:91]: arrayName is never null, Select() return a IEnumerable<out T>

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+
+                    var numberAsString = splitedName.Length != 2 && splitedName.Length >= 1
+                        ? splitedName[1]
+                        : null;
+
+                    var isArray = numberAsString?.Length >= 1
+                        ? numberAsString.Split(',')
+                        : null;
+
+                    if (isArray != null && isArray.Length == 1)
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:129]: Can be simplified

File: DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        /// <summary>
+        /// set the dimension as the name for each row
+        /// </summary>
+        private void SetVariableRowName()
+        {
+            foreach (var row in this.Variables)
+            {
+                row.Name = this.GetDimension();
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ArrayVariableRowViewModel.cs:177]: Multi calls on the same code

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs [hunk 31]
```diff
@@ -0,0 +1,170 @@
...
+        public ChooseMappingColumnsViewModel()
+        {
+        }
+
+        /// <summary>
+        /// ViewModel of the dialog to match columns of parameter to a variable
+        /// </summary>
+        public ChooseMappingColumnsViewModel(VariableBaseRowViewModel variable, ParameterOrOverrideBase parameter)
+        {
+
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs:57]: extra space

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs [hunk 32]
```diff
@@ -0,0 +1,170 @@
...
+                    groupby = array.Variables.GroupBy(x => this.GetColumnNumberFromValueName(x.Reference.DisplayName.Text)).ToList();
+                }
+
+                foreach (var group in groupby)
+                {
+                    this.ListOfVariableToMap.Add(new ChooseMappingRowsViewModel(group.ToList(), array.IsList));
+                }
+            }
+
+            this.ParameterName = ModelCode(parameter);
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs:85]: missing this, also why not using the Methods from the type itself:  parameter.ModelCode();

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs [hunk 33]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        /// <summary>
+        /// </summary>
+        /// Name of the variable to map to
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs:90]: misplaced doc

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs [hunk 34]
```diff
@@ -0,0 +1,170 @@
...
+
+        /// <summary>
+        /// List of value that will feed the dropdown of the dialog, from parameter
+        /// </summary>
+        public ReactiveList<string> ListOfParameterToMatch { get; set; } = new ReactiveList<string>();
+
+        /// <summary>
+        /// Interface to close the dialog
+        /// </summary>
+        public ICloseWindowBehavior CloseWindowBehavior { get; set; }
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs:116]: unused reference

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs [hunk 35]
```diff
@@ -0,0 +1,170 @@
...
+                    var num = splitedName.Length - 2;
+                    var a = ',' + splitedName[num];
+                    return a;
+                }
+            }
+
+            return null;
+        }
+
+
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs:142]: extra line

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs [hunk 36]
```diff
@@ -0,0 +1,170 @@
...
+        /// <summary>
+        /// Construct a Name from its parameters
+        /// </summary>
+        /// <returns>name of the element and his parameter type if available</returns>
+        private string ModelCode(ParameterOrOverrideBase parameter)
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs:147]: is it really needed?

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs [hunk 37]
```diff
@@ -0,0 +1,170 @@
...
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Lesser General Public License for more details.
+// 
+//    You should have received a copy of the GNU Lesser General Public License
+//    along with this program; if not, write to the Free Software Foundation,
+//    Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace DEHPEcosimPro.ViewModel.Rows
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingColumnsViewModel.cs:25]: Doesn't look like the right namespace for this dialog view model

File: DEHPEcosimPro/ViewModel/Rows/ChooseMappingRowsViewModel.cs [hunk 38]
```diff
@@ -0,0 +1,121 @@
...
+        /// <param name="isList">is the array a table with only one column, so a list ? like data[x] and not value[x,y]</param>
+        /// <returns>a string like [x] that represent the index of the column we want to map</returns>
+        private string GetIndexOfRow(List<VariableRowViewModel> list, bool isList = false)
+        {
+            if (isList)
+            {
+                return "[1]";
+            }
+
+            if (list.FirstOrDefault().Name.Contains('['))
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ChooseMappingRowsViewModel.cs:96]: possible null reference exception

File: DEHPEcosimPro/ViewModel/Rows/ValueSetsToTableViewModel.cs [hunk 39]
```diff
@@ -0,0 +1,205 @@
...
+        {
+        }
+
+        /// <summary>
+        /// Takes the valueset of a parameter and transform it into tables
+        /// </summary>
+        /// <param name="parameter">parameter that contain the valueset</param>
+        /// <param name="SelectedOption">option of the parameter to find the valueset</param>
+        /// <param name="SelectedState">state of the parameter to find the valueset</param>
+        public ValueSetsToTableViewModel(ParameterOrOverrideBase parameter, Option SelectedOption, ActualFiniteState SelectedState)
```
Comment (nathanatrhea) [inline DEHPEcosimPro/ViewModel/Rows/ValueSetsToTableViewModel.cs:54]: lower casing for parameter names

**Review by nathanatrhea — CHANGES_REQUESTED** [review]

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 40]
```diff
@@ -286,22 +326,56 @@ private void SelectedElementDefinitionChanged()
...
         public void AreVariableTypesAreCompatible()
         {
-            if (!(this.SelectedMappedElement?.SelectedVariable is {} variable && this.SelectedMappedElement.SelectedParameter is {} parameter))
+            if (!(this.SelectedAvailableVariables is {} variable && this.SelectedMappedElement?.SelectedParameter is {} parameter))
             {
                 return;
             }
+            
+            ValidationResult validationResult;
+            validationResult.Message = string.Empty;
```
[the remark this answers is not in the document]
Comment (arielleatrhea) [reply DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:335]: corrected

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs (hunk 5 above)
Comment (arielleatrhea) [reply DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:355]: used statusbar to avert the user

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 41]
```diff
@@ -286,22 +326,56 @@ private void SelectedElementDefinitionChanged()
...
+                        this.MapParameterToVariable(table.ListOfTuple, parameter);
+                    }
+                }
+                else
+                {
+                    //not compatible, show error to user
+                }
 
-            var validationResult = parameter.ParameterType.Validate(variable.ActualValue, parameter.Scale);
+                validationResult.ResultKind = userSelectecValue == true && areArraysCompatible
```
[the remark this answers is not in the document]
Comment (arielleatrhea) [reply DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:358]: not in the case where rows are different (i did unit test)

File: DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs [hunk 42]
```diff
@@ -313,21 +387,76 @@ private void SelectedThingChanged()
             switch (this.SelectedThing)
             {
                 case IRowViewModelBase<ElementDefinition> elementDefinition:
```
[the remark this answers is not in the document]
Comment (arielleatrhea) [reply DEHPEcosimPro/ViewModel/Dialogs/HubMappingConfigurationDialogViewModel.cs:389]: i have to do more test to be sure

**Pull request merged**